src/storage/txn/latch.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+use std::collections::{VecDeque, BTreeSet};
+use std::hash::{Hash, SipHasher, Hasher};
+
+// simulate lock of one row
+struct Latch {
+    // store waiting commands
+    pub waiting: VecDeque<u64>,
+
+    // use to check existed
+    pub set: BTreeSet<u64>,
+}
+
+impl Latch {
+    pub fn new() -> Latch {
+        Latch {
+            waiting: VecDeque::new(),
+            set: BTreeSet::new(),
+        }
+    }
+}
+
+pub struct Latches {
+    locks: Vec<Latch>,
+    size: usize,
+}
+
+impl Latches {
+    pub fn new(size: usize) -> Latches {
+        Latches {
+            locks: (0..size).map(|_|  Latch::new()).collect(),
+            size: size,
+        }
+    }
+
+    pub fn calc_latches_indices<H>(&self, keys: &[H]) -> Vec<usize>
+        where H: Hash
+    {
+        let mut indices: Vec<usize> = keys.iter().map(|x| self.calc_index(x)).collect();
+
+        // prevent from deadlock, so we sort and deduplicate the index
+        indices.sort();
+        indices.dedup();
+        indices
+    }
+
+    pub fn acquire_by_indices(&mut self, indices: &[usize], who: u64) -> usize {
+
+        let mut acquired_count: usize = 0;
+        for i in indices {
+            let rowlock = &mut self.locks.get_mut(*i).unwrap();
+
+            let mut front: Option<u64> = None;
+            if let Some(cid) = rowlock.waiting.front() {
+                front = Some(*cid);
+            }
+
+            match front {
+                Some(cid) => {
+                    if cid == who {
+                        acquired_count += 1;
```

Comment:
Looks not possible to happen. When you call `acquire_by_indices`, you only pass the indices that not acquired yet.